README
This directory contains a set of network tests and demonstrations
programs.  

    netdemo  - simplest test, echos data on sockets
    dnstest  - test of DNS client
    http     - test of GoAhead web server
    netlink  - link only test to evaluate code space requirement
               of network stack, ftpd, and httpd
    ntp      - test of network time protocol client
    rpc_demo - test of RPC library
    select   - test of select() routine functionality
    tftpTest - test of TFTP (Trivial FTP) client filesystem 
    ttcp     - TTCP throughput test

The file networkconfig.h MUST be configured to reflect your local
network according to the included Configuration Notes.

CONFIGURATION NOTES
===================

1. Make sure you change the definition of MY_ETHERNET_ADDRESS in 
   networkconfig.h before you try to use any of the network
   demonstration programs.

2. If you do not use BOOTP (i.e. the `#define USE_BOOTP 1' in init.c is
   removed), make sure you change the definition of MY_INTERNET_ADDRESS
   before you try to use the programs.

FAILURE TO FOLLOW THE ABOVE INSTRUCTIONS MAY PROVOKE YOUR NETWORK
ADMINISTRATOR TO THE POINT WHERE VICIOUS ATTACKS WILL BE MADE UPON YOU.



ChangeLog
2011-08-02	Joel Sherrill <joel.sherrilL@OARcorp.com>

	* Makefile: Use new path.

2011-03-29     Sebastian Huber <Sebastian.Huber@embedded-brains.de>

	* mcast/listener.c: Output only received characters.

2011-03-03	Joel Sherrill <joel.sherrill@oarcorp.com>

	* networkconfig-qemu.h: VirtualBox uses 10.0.x.x for its eth0 so we
	need to switch to 10.1.x.x for things to work smoothly.
	* networkconfig-loopback.h: New file.

2011-03-03	Joel Sherrill <joel.sherrill@oarcorp.com>

	* networkconfig-qemu.h: Add missing endif.

2010-12-08     Sebastian Huber <Sebastian.Huber@embedded-brains.de>

	* networkconfig.h, networkconfig-qemu.h, networkconfig-icecube.h,
	networkconfig-mpc8313erdb.h: Loop back interface will be initialized
	now during normal initialization via rtems_bsdnet_initialize_network().

2010-02-12	Joel Sherrill <joel.sherrill@oarcorp.com>

	* ttcp/rtems_ttcp.c: Update timeval usage.

2009-11-13	Joel Sherrill <joel.sherrill@OARcorp.com>

	* VERSION: Updated to network-demos-4.9.3.

2009-05-13	Joel Sherrill <joel.sherrill@oarcorp.com>

	* networkconfig-qemu.h: Now works on qemu with fxp driver in RTEMS, the
	newer version in libbsdport, and on the real NIC in the Winsystems PC
	in RTEMS Lab.

2009-02-10	Joel Sherrill <joel.sherrill@oarcorp.com>

	* networkconfig-qemu.h: Add option for qemu to use PCI i82551 NIC.

2008-09-18	Joel Sherrill <joel.sherrill@OARcorp.com>

	* networkconfig.h: Revert previous patch. Use loopback configuration
	only.

2008-09-17	Joel Sherrill <joel.sherrill@oarcorp.com>

	* networkconfig-mpc8313erdb.h, networkconfig.h: Rename STACK_CHECKER_ON
	to more appropriate CONFIGURE_STACK_CHECKER_ENABLED.

2008-08-21	Joel Sherrill <joel.sherrill@oarcorp.com>

	* networkconfig-mpc8313erdb.h: New file.

2008-08-20	Joel Sherrill <joel.sherrill@OARcorp.com>

	* networkconfig-qemu.h, networkconfig.h: Add missing initializers.

2008-02-05	Joel Sherrill <joel.sherrill@oarcorp.com>

	* networkconfig-icecube.h, networkconfig-qemu.h, networkconfig.h:
	Updated to include new fields and expand tabs.

2007-09-25	Joel Sherrill <joel.sherrill@oarcorp.com>

	* Makefile: Added telnetd test.

2007-09-25	Joel Sherrill <joel.sherrill@oarcorp.com>

	* networkconfig-icecube.h: Fix MAC address and typo.

2007-09-19	Joel Sherrill <joel.sherrill@oarcorp.com>

	* Makefile: Add mcast.

2007-09-14	Joel Sherrill <joel.sherrill@oarcorp.com>

	* networkconfig.h: Loopback only network configuration tweaked to have
	a hostname.

2007-09-14	Joel Sherrill <joel.sherrill@oarcorp.com>

	* networkconfig.h: Loopback only network configuration. Use one of the
	real ones as an example.

2007-09-07	Joel Sherrill <joel.sherrill@oarcorp.com>

	* networkconfig-qemu.h: Disable DHCP and use static IP.
	* networkconfig.h: Use only loopback so tests can run OK on any target
	even if they can't do anything.
	* networkconfig-icecube.h: New file.

2007-08-07	Joel Sherrill <joel.sherrill@OARcorp.com>

	* VERSION: Updated to network-demos-4.7.99.2.

2007-07-18	Joel Sherrill <joel.sherrill@oarcorp.com>

	* networkconfig-qemu.h: New file.

2007-07-18	Joel Sherrill <joel.sherrill@oarcorp.com>

	* networkconfig.h: Make the NFS server, remote filesystem, and
	directory to look at configurable.

2007-06-22	Joel Sherrill <joel.sherrill@OARcorp.com>

	* Makefile: Remove netlink. It appeared to be a hacked copy of http.
	* netlink/ChangeLog, netlink/Makefile, netlink/index.html,
	netlink/init.c, netlink/networkconfig.h, netlink/system.h: Removed.

2007-06-21	Joel Sherrill <joel.sherrill@OARcorp.com>

	* Makefile: New file.

2007-03-13	Joel Sherrill <joel@OARcorp.com>

	* VERSION: Updated to network-demos-4.7.99.1.

2005-09-02	Joel Sherrill <joel@OARcorp.com>

	* VERSION: Updated to network-demos-4.6.99.2.

2003-09-18	Joel Sherrill <joel@OARcorp.com>

	PR 493/networking
	* dnstest/init.c, http/Makefile, http/init.c, http/system.h,
	netdemo/init.c, netlink/init.c, ntp/init.c, select/init.c,
	select/test.c, tftpTest/init.c, tftpTest/test.c, ttcp/init.c,
	ttcp/rtems_ttcp.c: Made all the tests compile and eliminated warnings
	except in the ttcp_orig code.
	* Makefile: New file.
	* netlink/system.h: Removed.

2003-07-03	Joel Sherrill <joel@OARcorp.com>

	* VERSION: Updated to network-demos-4.6.0pre4.

2003-02-11	Joel Sherrill <joel@OARcorp.com>

	* VERSION: Updated to network-demos-4.6.0pre1.

2003-09-18	Jay Monkman <jtm@smoothsmoothie.com>

	PR 494/networking
	* Changed Makefile to not use objcopy to create a linkable
        object file from the tar file. Instead, ld links it in
        directly.

2003-09-18	Joel Sherrill <joel@OARcorp.com>

	PR 493/networking
	* dnstest/init.c, http/Makefile, http/init.c, http/system.h,
	netdemo/init.c, netlink/init.c, ntp/init.c, select/init.c,
	select/test.c, tftpTest/init.c, tftpTest/test.c, ttcp/init.c,
	ttcp/rtems_ttcp.c: Made all the tests compile and eliminated warnings
	except in the ttcp_orig code.
	* Makefile: New file.
	* netlink/system.h: Removed.

2003-07-03	Joel Sherrill <joel@OARcorp.com>

	* VERSION: Updated to network-demos-4.6.0pre4.

2003-02-11	Joel Sherrill <joel@OARcorp.com>

	* VERSION: Updated to network-demos-4.6.0pre1.

2003-01-28	Joel Sherrill <joel@OARcorp.com>

	* VERSION: Updated to network-demos-20030128.

2002-10-21	Joel Sherrill <joel@OARcorp.com>

	* VERSION: Updated to network-demos-20021018.

2002-10-18	Joel Sherrill <joel@OARcorp.com>

	* dnstest/init.c, netdemo/init.c, netlink/init.c, ntp/init.c,
	rpc_demo/init.c, select/init.c, tftpTest/init.c, ttcp/init.c:
	define CONFIGURE_LIBIO_MAXIMUM_FILE_DESCRIPTORS to 20 for enough
	sockets and CONFIGURE_USE_IMFS_AS_BASE_FILESYSTEM even if not
	absolutely needed for networking.

2001-08-30	Joel Sherrill <joel@OARcorp.com>

	* SUPPORT: New file.

2001-08-30	Joel Sherrill <joel@OARcorp.com>

	* README, networkconfig.h, netdemo/README, tftpTest/test.c: Merged
	missing stuff from 4.5 branch.
	* http/Makefile, http/init.c: Minor modifications to improve
	buildability.

2000-12-14	Eric Norum <eric.norum@usask.ca>


http/init.c
/*  Init
 *
 *  This routine is the initialization task for this test program.
 *
 *  Don't forget to change the IP addresses
 */

#include "system.h"
#include <bsp.h>

#include <errno.h>
#include <time.h>

#include <stdio.h>
#include <rtems/rtems_bsdnet.h>
#include <rtems/ftpd.h>
#include <rtems/untar.h>

     
#include <rtems/error.h>
#include <rpc/rpc.h>
#include <netinet/in.h>
#include <time.h>

#include <arpa/inet.h>
#include <sys/socket.h>
#include "../networkconfig.h"

#include <rtems_webserver.h>

#define ARGUMENT 0

/*
 *  The tarfile image is built automatically externally.
 */
#include "FilesystemImage.h"

#if defined(USE_FTPD)
  bool FTPD_enabled = true;
  struct rtems_ftpd_configuration rtems_ftpd_configuration = {
    10,                     /* FTPD task priority            */
    1024,                   /* Maximum buffersize for hooks  */
    21,                     /* Well-known port     */
    NULL,                   /* List of hooks       */
    NULL,                   /* Root for FTPD or 0 for / */
    0,                      /* Max. connections    */
    0,                      /* Idle timeout in seoconds
                               or 0 for no (inf) timeout */
    0,                      /* 0 - r/w, 1 - read-only,
                               2 - write-only,
                               3 - browse-only */
 };

#else
 bool FTPD_enabled = false;
#endif

#if defined(USE_GOAHEAD_HTTPD) && !defined(RTEMS_POSIX_API)
  #warning "GoAhead server requires POSIX API - switching to SHTTPD"
  #undef USE_GOAHEAD_HTTPD
  #undef USE_SIMPLE_HTTPD
#endif

#if defined(USE_GOAHEAD_HTTPD)
  bool GoAhead_HTTPD_enabled = true;

  /* GoAhead Trace Handler */
  #include <goahead/uemf.h>
  void quietTraceHandler(int level, char *buf)
  {
    /* do nothing */
  }
#else
  bool GoAhead_HTTPD_enabled = false;
#endif

#if defined(USE_SIMPLE_HTTPD)
  bool Simple_HTTPD_enabled = true;

  #include <shttpd/shttpd.h>
#else
  bool Simple_HTTPD_enabled = false;
#endif

#if defined(USE_MONGOOSE_HTTPD)
  #include <mghttpd/mongoose.h>

  void example_mongoose_addpages(
    struct mg_context *server
  );
#endif

#define bool2string(_b) ((_b) ? "true" : "false")

#if defined(USE_SIMPLE_HTTPD)
extern void example_shttpd_addpages(struct shttpd_ctx *ctx);
#endif

rtems_task Init(
  rtems_task_argument argument
)
{
  rtems_status_code status;

  printf("\n\n*** HTTP TEST ***\n\r" );
  printf("GoAhead HTTPD Enabled: %s\n", bool2string(GoAhead_HTTPD_enabled) );
  printf("Simple HTTPD Enabled: %s\n", bool2string(Simple_HTTPD_enabled) );
  printf("FTPD Enabled: %s\n", bool2string(FTPD_enabled) );
  printf("\n");

  /*
   * Load filesystem image
   */
  printf("Loading filesystem image\n");
  status = Untar_FromMemory( (char *)FilesystemImage, FilesystemImage_size );
   
  printf("Initializing Network\n");
  rtems_bsdnet_initialize_network ();

  #if defined(USE_FTPD)
    printf( "Initializing FTPD\n" );
    rtems_initialize_ftpd();
  #endif

  #if defined(USE_GOAHEAD_HTTPD)
    printf( "Initializing GoAhead HTTPD\n" );
    status = rtems_initialize_webserver();
    if ( status )
      printf( "ERROR -- failed to initialize webserver\n" );

    traceSetHandler( quietTraceHandler );
  #endif

  #if defined(USE_SIMPLE_HTTPD)
    /*
     *  SHTTPD uses about 37K of stack even in this test on a PowerPC.
     *  There is no point in doing math for the stack size.  Bump it
     *  until there isn't a problem.
     */
    printf( "Initializing Simple HTTPD\n" );
    status = rtems_initialize_webserver(
      100,                             /* initial priority */
      (48 * 1024),                     /* stack size */
      RTEMS_DEFAULT_MODES,             /* initial modes */
      RTEMS_DEFAULT_ATTRIBUTES,        /* attributes */
      NULL,                            /* init_callback */
      example_shttpd_addpages,         /* addpages_callback */
      "/",                             /* initial priority */
      80                               /* port to listen on */
    );
    if ( status )
      printf( "ERROR -- failed to initialize webserver\n" );

  #endif

  #if defined(USE_MONGOOSE_HTTPD)
    {
      struct mg_context *WebServer;
      printf( "Initializing Mongoose HTTPD\n" );
      WebServer = mg_start();

      mg_set_option(WebServer, "root", "/" );
      mg_set_option(WebServer, "ports", "80" );
      example_mongoose_addpages( WebServer );

    }
  #endif
  status = rtems_task_delete( RTEMS_SELF );
}


/*
 *  Configuration
 */
#define CONFIGURE_APPLICATION_NEEDS_CONSOLE_DRIVER
#define CONFIGURE_APPLICATION_NEEDS_CLOCK_DRIVER
#define CONFIGURE_RTEMS_INIT_TASKS_TABLE
#define CONFIGURE_LIBIO_MAXIMUM_FILE_DESCRIPTORS	20
#define CONFIGURE_USE_IMFS_AS_BASE_FILESYSTEM

#define CONFIGURE_EXECUTIVE_RAM_SIZE	(512*1024)
#define CONFIGURE_MAXIMUM_SEMAPHORES	20
#define CONFIGURE_MAXIMUM_TASKS		20

#if defined(USE_MONGOOSE_HTTPD)
#define CONFIGURE_MAXIMUM_POSIX_THREADS 10
#define CONFIGURE_MAXIMUM_POSIX_MUTEXES 30
#define CONFIGURE_MAXIMUM_POSIX_CONDITION_VARIABLES 10
#define CONFIGURE_MINIMUM_TASK_STACK_SIZE (32 * 1024)
#define CONFIGURE_UNIFIED_WORK_AREAS
#endif

#define CONFIGURE_MICROSECONDS_PER_TICK	10000

#define CONFIGURE_INIT_TASK_STACK_SIZE	(10*1024)
#define CONFIGURE_INIT_TASK_PRIORITY	120
#define CONFIGURE_INIT_TASK_INITIAL_MODES (RTEMS_PREEMPT | \
                                           RTEMS_NO_TIMESLICE | \
                                           RTEMS_NO_ASR | \
                                           RTEMS_INTERRUPT_LEVEL(0))

#define CONFIGURE_STACK_CHECKER_ENABLED
#define CONFIGURE_INIT

#include <rtems/confdefs.h>

